### PR TITLE
Sequence annotation restart

### PR DESCRIPTION
**Pull request by macieksmuga**
Try 2. Replaces #962

### CODE REVIEW

**Comment by macieksmuga** [comment]
@dcolligan @saupchurch @jeromekelleher: This PR is now complete modulo some simulated and data-driven tests, which I'm currently in the middle of writing. The functionality is all there, so please review with the goal of getting it merged in ASAP.

File: ga4gh/_protocol_definitions.py [hunk 1]
```diff
@@ -207,6 +207,52 @@ def __init__(self, **kwargs):
         """
 
 
+class Attributes(ProtocolElement):
+    """
+    Type defining a collection of attributes associated with various
+    protocol   records.  Each attribute is a name that maps to an
+    array of one or more   values.  Values can be strings, external
+    identifiers, or ontology terms.   Values should be split into the
```
Comment (dcolligan) [inline ga4gh/_protocol_definitions.py]: This isn't the biggest deal in the world, but it seems like there are some weird double-spaces in these comments.  Is `process_schemas.py` inserting extra spaces?

File: ga4gh/backend.py [hunk 2]
```diff
@@ -565,6 +565,44 @@ def variantAnnotationsGenerator(self, request):
...
+            parentId = compoundParentId.featureId
+            # A client can optionally specify JUST the (compound) parentID,
+            # and the server needs to derive the dataset & featureSet
+            # from this (compound) parentID.
+            if compoundId is None:
+                compoundId = compoundParentId
+            else:
+                # check that the dataset and featureSet of the parent
+                # compound ID is the same as that of the featureSetId
+                if compoundParentId.datasetId \
```
Comment (dcolligan) [inline ga4gh/backend.py]: surround this condition with parens instead of using \'s

File: ga4gh/cli.py [hunk 3]
```diff
@@ -413,6 +442,21 @@ def _textOutput(self, gaObjects):
...
+    Mix-in class to format Feature (Sequence Annotation) objects
+    """
+    def _textOutput(self, gaObjects):
+        for feature in gaObjects:
+            print(
+                feature.id, feature.parentId, feature.featureSetId,
+                feature.referenceName, feature.start, feature.end,
+                feature.strand, sep="\t", end="\t")
+            # TODO: handle attributes
+            # TODO: handle featureType (OntologyTerm)
```
Comment (dcolligan) [inline ga4gh/cli.py]: are these going to be completed?

File: ga4gh/client.py [hunk 4]
```diff
@@ -292,6 +310,36 @@ def searchVariantAnnotations(
...
+        """
+        request = protocol.SearchFeaturesRequest()
+        request.featureSetId = featureSetId
+        request.parentId = parentId
+        request.referenceName = referenceName
+        request.start = start
+        request.end = end
+        request.ontologyTerms = ontologyTerms
+        request.pageSize = self._pageSize
+        print(request)
```
Comment (dcolligan) [inline ga4gh/client.py]: forgot to take this out?

File: ga4gh/client.py [hunk 5]
```diff
@@ -337,6 +385,22 @@ def searchVariantAnnotationSets(self, variantSetId):
...
+        :param str datasetId: The ID of the
+            :class:`ga4gh.protocol.Dataset` of interest.
+        :return: An iterator over the :class:`ga4gh.protocol.FeatureSet`
+            objects defined by the query parameters.
+        """
+        request = protocol.SearchFeatureSetsRequest()
+        request.datasetId = datasetId
+        request.pageSize = self._pageSize
+        return self._runSearchRequest(
+                    request, "featuresets", protocol.SearchFeatureSetsResponse)
```
Comment (dcolligan) [inline ga4gh/client.py]: this is indented more than it needs to be

File: ga4gh/datamodel/ontologies.py [hunk 6]
```diff
@@ -7,6 +7,7 @@
 
 import fnmatch
 import os
+import ga4gh.protocol as protocol
 
 
 class FileSystemOntology(object):
```
Comment (dcolligan) [inline ga4gh/datamodel/ontologies.py]: need abstract/simulated ontology

File: ga4gh/datamodel/sequenceAnnotations.py [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+        except without the pagetoken/size.
+        """
+        sql = ("SELECT COUNT(*) FROM FEATURE WHERE "
+               "reference_name = ? "
+               "AND end > ? "  # compare this to query start
+               "AND start < ? "  # and this to query end
+               )
+        sql_args = (referenceName, start, end)
+        if parentId is not None:
+            sql += "AND parent_id = ? "
```
Comment (dcolligan) [inline ga4gh/datamodel/sequenceAnnotations.py]: There's got to be a better way of doing this than string concatenation.   Do this instead: put formatting arguments in the `sql` string and set them to empty strings if not applicable.  e.g.

```
limit = ''
sql = 'SELECT * FROM Table {}'
if condition:
    limit = 'LIMIT 10'
finalSql = sql.format(limit)
```

Comment (dcolligan) [inline ga4gh/datamodel/sequenceAnnotations.py]: (I would note that format also accepts dicts... check out that syntax)

File: ga4gh/datamodel/sequenceAnnotations.py [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+        :return an array of dictionaries, representing the returned data.
+        """
+        sql = ("SELECT * FROM FEATURE WHERE "
+               "reference_name = ? "
+               "AND end > ? "  # compare this to query start
+               "AND start < ? "  # and this to query end
+               )
+        sql_args = (referenceName, start, end)
+        if parentId is not None:
+            sql += "AND parent_id = ? "
```
Comment (dcolligan) [inline ga4gh/datamodel/sequenceAnnotations.py]: same comment about string concatenation here

File: ga4gh/datamodel/sequenceAnnotations.py [hunk 9]
```diff
@@ -0,0 +1,301 @@
...
+        :param featureId: string feature ID as reported in GFF3
+        :return: string representing ID for the specified GA4GH protocol
+            Feature object in this FeatureSet.
+        """
+        compoundId = datamodel.FeatureCompoundId(
+            self.getCompoundId(), _decolonize(featureId))
+        return str(compoundId)
+
+
+class SimulatedFeatureSet(AbstractFeatureSet):
```
Comment (dcolligan) [inline ga4gh/datamodel/sequenceAnnotations.py]: needs doc comment

File: ga4gh/datamodel/sequenceAnnotations.py [hunk 10]
```diff
@@ -0,0 +1,301 @@
...
+    def __init__(self, parentContainer, localId, filePath, dataRepository):
+        super(Gff3DbFeatureSet, self).__init__(
+            parentContainer, localId, None)
+        self._sequenceOntology = dataRepository.getOntology(
+            'sequence_ontology')
+        self._dbFilePath = filePath
+        self._dataRepository = dataRepository
+        self._db = Gff3DbBackend(self._dbFilePath)
+
+    def getFeature(self, compoundId):
```
Comment (dcolligan) [inline ga4gh/datamodel/sequenceAnnotations.py]: doc comment

File: ga4gh/datamodel/sequenceAnnotations.py [hunk 11]
```diff
@@ -0,0 +1,301 @@
...
+        if feature.get('parent_id'):
+            gaFeature.parentId = self.getCompoundIdForFeatureId(
+                    feature['parent_id'])
+        else:
+            gaFeature.parentId = None
+        gaFeature.featureSetId = self.getId()
+        gaFeature.referenceName = feature['reference_name']
+        gaFeature.start = int(feature['start'])
+        gaFeature.end = int(feature['end'])
+        print(feature['child_ids'])  # DEBUG
```
Comment (dcolligan) [inline ga4gh/datamodel/sequenceAnnotations.py]: take out

Comment (dcolligan) [inline ga4gh/datamodel/sequenceAnnotations.py]: ps you know about import pdb; pdb.set_trace(), right?

File: ga4gh/datamodel/sequenceAnnotations.py [hunk 12]
```diff
@@ -0,0 +1,301 @@
...
+        gaFeature.childIds = map(
+                self.getCompoundIdForFeatureId,
```
Comment (dcolligan) [inline ga4gh/datamodel/sequenceAnnotations.py]: less indent here

File: ga4gh/gff3Parser.py [hunk 13]
```diff
@@ -0,0 +1,321 @@
...
+    """
+    Exception associated with GFF3 data.
+    """
+    def __init__(self, message, fileName=None, lineNumber=None):
+        if fileName is not None:
+            message = "{}:{}: {}".format(fileName, lineNumber, message)
+        super(GFF3Exception, self).__init__(message)
+
+# characters forcing encode for columns
+_encodeColReStr = "\t|\n|\r|%|[\x00-\x1F]|\x7f"
```
Comment (dcolligan) [inline ga4gh/gff3Parser.py]: it's easier to read this variable name without acronyms... change 'Re' to 'Regex' ... same for the vars below

File: ga4gh/serverconfig.py [hunk 14]
```diff
@@ -40,6 +40,8 @@ class DevelopmentConfig(BaseConfig):
     """
     DATA_SOURCE = "ga4gh-example-data"
     DEBUG = True
+    REQUEST_VALIDATION = False
```
Comment (dcolligan) [inline ga4gh/serverconfig.py]: Why are we turning off validation here?

**Comment by dcolligan** [comment]
This is... a lot of code.  A few thoughts:
- I didn't see any tests for 'sqliteBackend.py' ... that seems like something that needs tests
- We should make sure the scripts for creating example data repos still work after your addition (which adds a new top-level directory) with the server
- I haven't been working on SA stuff, so there's a lot in here that I don't understand.  Perhaps we can talk about it at some point in the near future.

File: ga4gh/gff3Parser.py (hunk 13 above)
Comment (diekhans) [inline ga4gh/gff3Parser.py]: you mean with a different acronym :-)

File: ga4gh/datamodel/sequenceAnnotations.py (hunk 7 above)
Comment (macieksmuga) [inline ga4gh/datamodel/sequenceAnnotations.py]: I'd totally agree with you if this were inside the body of a tighter loop, but it's done once per request - can we leave it like this for now (it's logically correct), then optimize for performance once we have some realistic timing metrics?

**Comment by macieksmuga** [comment]
Thanks @dcolligan for the review! 

I'm making a lot of the fixes you suggested right now, you should see an update by the end of the day. And yeah, it's a lot of code: I needed to provide quite a bit of brand-new functionality to the server, not the least of which was hooking it up to a completely new SQLite-driven backend. As for your specific points:
* Yeah, `sqliteBackend` unit tests would be a good idea. Thanks for reminding me :smile:
* I did try to make sure the dataRepo tools work on the new `sequenceAnnotations` directory. If you're thinking of the `ontologies` directory, that came in previously with Variation Annotations, I should just be inheriting the code that's already there.
* I'll be happy to walk you and Jerome through it!

File: ga4gh/_protocol_definitions.py (hunk 1 above)
Comment (macieksmuga) [inline ga4gh/_protocol_definitions.py]: It's all autogenerated code, must be an issue with the underlying schemas documentation, thanks.

File: ga4gh/datamodel/sequenceAnnotations.py (hunk 7 above)
Comment (dcolligan) [inline ga4gh/datamodel/sequenceAnnotations.py]: ok

File: ga4gh/_protocol_definitions.py (hunk 1 above)
Comment (dcolligan) [inline ga4gh/_protocol_definitions.py]: I created a low-priority issue to track it; it's not something you introduced https://github.com/ga4gh/server/issues/993

File: ga4gh/datamodel/ontologies.py (hunk 6 above)
Comment (macieksmuga) [inline ga4gh/datamodel/ontologies.py]: Agree, but not on this PR.

**Comment by macieksmuga** [comment]
I'd like to supersede this PR with #999, which is fully equivalent to this one, but squashes all the linear commits and rebases properly against the latest master.

**Comment by macieksmuga** [comment]
Closing this PR for now in favor of #999. Will keep the branch behind this around in case anybody wants to reference the detailed development history for sequence annotations.

**Pull request closed without merging**